### ISSUE TITLE
Modularizing OSHI

### ISSUE THREAD

**Issue by dbwiddis** [dblock/oshi#200]
First issue in a series of API-breaking changes for OSHI 3.0.

The JSON functionality of OSHI should be separated into a separate module.  Reasoning:
* The JSON functionality requires added dependencies which are unnecessary for those who do not need that feature. The core class would return to dependencies only on JNA and slf4j-api.
* The functionality is easily abstracted into separate code that need not be installed by users who do not need it (simple copy/paste of the `toJSON()` methods.)
* The same functionality could easily be extended to XML or other concepts, which all belong in separate modules leveraging the core code, rather than adding additional dependencies and footprint to the existing code.

Overall concept:
* Existing `oshi-core` module would remain as-is except for the inheritance from the `OshiJSONObject` class which specifies the `toJSON()` method.  
* New `oshi-json` module would be created with decorator classes for the OSHI API (`SystemInfo` and the classes in `oshi.hardware` and `oshi.software.os`.  These would be in the package `oshi.json.*` rather than `oshi.*`, and would internally instantiate a private object from the core class, providing the necessary access methods (implementing the interface) to the core object, plus adding the `toJSON()` functionality.
* Consider adding configurability to the `oshi-json` class using Apache Commons Configuration to fine tune what information is returned in the JSON string.  (Default: everything.) (#208)
* Consider switching from `javax.json` to a more flexible/widespread/capable/actively supported JSON provider (Jackson?) (#206)

Questions:
* `oshi-core` will have to move to 3.0.  Should `oshi-json` match version numbers or start at 1.0?  Will both versions always be released/incremented together or maintain a separate release schedule and only increment when necessary?
* will @hazendaz do the heavy lifting of rewriting the `pom.xml` files to make this work, including easy publishing of artifacts to the Central Repository and documents to the gh-pages branch?
* is JSON too narrow of a focus?  Should we provide both JSON and xml in a single second project or have two (or more) separate projects?  What widespread db formats (Mongo?) would use it and should we specifically design output in an object easily ingested into those dbs?

**Comment by hazendaz** [comment]
All modules should use same version numbers.  Spring framework does it that way.  Its far easier for releasing even if each piece not touched.

Get Outlook for Android<https://aka.ms/ghei36>



On Mon, Jun 20, 2016 at 12:32 AM -0400, "dbwiddis" <notifications@github.com<mailto:notifications@github.com>> wrote:


First issue in a series of API-breaking changes for OSHI 3.0.

The JSON functionality of OSHI should be separated into a separate module. Reasoning:

  *   The JSON functionality requires added dependencies which are unnecessary for those who do not need that feature. The core class would return to dependencies only on JNA and slf4j-api.
  *   The functionality is easily abstracted into separate code that need not be installed by users who do not need it (simple copy/paste of the toJSON() methods.)
  *   The same functionality could easily be extended to XML or other concepts, which all belong in separate modules leveraging the core code, rather than adding additional dependencies and footprint to the existing code.

Overall concept:

  *   Existing oshi-core module would remain as-is except for the inheritance from the OshiJSONObject class which specifies the toJSON() method.
  *   New oshi-json module would be created with decorator classes for the OSHI API (SystemInfo and the classes in oshi.hardware and oshi.software.os. These would be in the package oshi.json.* rather than oshi.*, and would internally instantiate a private object from the core class, providing the necessary access methods (implementing the interface) to the core object, plus adding the toJSON() functionality.
  *   Consider adding configurability to the oshi-json class using Apache Commons Configuration to fine tune what information is returned in the JSON string. (Default: everything.)
  *   Consider switching from javax.json to a more flexible/widespread/capable/actively supported JSON provider (Jackson?)

Questions:

  *   oshi-core will have to move to 3.0. Should oshi-json match version numbers or start at 1.0? Will both versions always be released/incremented together or maintain a separate release schedule and only increment when necessary?
  *   will @hazendaz<https://github.com/hazendaz> do the heavy lifting of rewriting the pom.xml files to make this work, including easy publishing of artifacts to the Central Repository and documents to the gh-pages branch?
  *   is JSON too narrow of a focus? Should we provide both JSON and xml in a single second project or have two (or more) separate projects? What widespread db formats (Mongo?) would use it and should we specifically design output in an object easily ingested into those dbs?

-

**Comment by dblock** [comment]
I vote for keeping everything in this project, but splitting JARs as suggested, releasing everything together and keeping version numbers consistent. If one day there's a community of hundreds of developers working on Oshi JSON formatting then maybe it deserves its own project ;)

**Comment by dbwiddis** [comment]
Sounds like a plan and agreement.   @hazendaz what's the best sequence for this to happen?  (And I need to release 2.6.2 with the latest tweaks before we start messing with things.)   Should I reorganize folders (tell me what structure to use) and then you come behind with the cleanup of the other files?  Or do you want to set up the new structure and then have me move the src files into spots you've created?

**Comment by dbwiddis** [comment]
OK, #205 separates the two projects into subdirectories `oshi-core` and `oshi-json`.  Everything works when I run it locally.

I haven't yet merged it but it's ready to merge. (Yes, Travis-CI is failing because updating travis is part of this task that I need help with!)

Copied the parent pom.xml to both subprojects with following changes:
 - removed json dependency from oshi-core project
 - removed jna dependency from oshi-json project; added oshi-core dependency

A few questions/concerns:
 - Both poms reference git-commit-id plugin and look for the .git directory in the project base.  I couldn't figure out how to make that work with the parent directory so I symlinked my (one) .git directory into both projects.  I hope that doesn't cause any problems... please fix if it does cause problems.
 - Would be nice to always have the oshi-json dependency on the latest/same version of oshi-core.  I presume there is a way to easily do that.

@hazendaz please do your magic!

**Comment by hazendaz** [comment]
Ill jump on tonight and sprinkle on a little magic.  Sounds close...

Get Outlook for Android<https://aka.ms/ghei36>



On Wed, Jun 22, 2016 at 6:01 PM -0400, "dbwiddis" <notifications@github.com<mailto:notifications@github.com>> wrote:


OK, #205<https://github.com/dblock/oshi/pull/205> separates the two projects into subdirectories oshi-core and oshi-json. Everything works when I run it locally.

I haven't yet merged it but it's ready to merge. (Yes, Travis-CI is failing because updating travis is part of this task that I need help with!)

Copied the parent pom.xml to both subprojects with following changes:

  *   removed json dependency from oshi-core project
  *   removed jna dependency from oshi-json project; added oshi-core dependency

A few questions/concerns:

  *   Both poms reference git-commit-id plugin and look for the .git directory in the project base. I couldn't figure out how to make that work with the parent directory so I symlinked my (one) .git directory into both projects. I hope that doesn't cause any problems... please fix if it does cause problems.
  *   Would be nice to always have the oshi-json dependency on the latest/same version of oshi-core. I presume there is a way to easily do that.

@hazendaz<https://github.com/hazendaz> please do your magic!

-

**Comment by hazendaz** [comment]
Sorry didnt get time tonight.  Ill try again tomorrow night to take a look.

Btw...if im reading right here maven inherites so you will want the build plugins all in the parent.  If you got that...cool..if not try to switch that over.  Parent will be in the root and travis will just work ;)

Get Outlook for Android<https://aka.ms/ghei36>



On Wed, Jun 22, 2016 at 6:09 PM -0400, "Jeremy Landis" <jeremylandis@hotmail.com<mailto:jeremylandis@hotmail.com>> wrote:


Ill jump on tonight and sprinkle on a little magic.  Sounds close...

Get Outlook for Android<https://aka.ms/ghei36>



On Wed, Jun 22, 2016 at 6:01 PM -0400, "dbwiddis" <notifications@github.com<mailto:notifications@github.com>> wrote:


OK, #205<https://github.com/dblock/oshi/pull/205> separates the two projects into subdirectories oshi-core and oshi-json. Everything works when I run it locally.

I haven't yet merged it but it's ready to merge. (Yes, Travis-CI is failing because updating travis is part of this task that I need help with!)

Copied the parent pom.xml to both subprojects with following changes:

  *   removed json dependency from oshi-core project
  *   removed jna dependency from oshi-json project; added oshi-core dependency

A few questions/concerns:

  *   Both poms reference git-commit-id plugin and look for the .git directory in the project base. I couldn't figure out how to make that work with the parent directory so I symlinked my (one) .git directory into both projects. I hope that doesn't cause any problems... please fix if it does cause problems.
  *   Would be nice to always have the oshi-json dependency on the latest/same version of oshi-core. I presume there is a way to easily do that.

@hazendaz<https://github.com/hazendaz> please do your magic!

-

**Comment by dbwiddis** [comment]
OK, I took a peek at waffle to try to grok the pattern and I'll give it a shot.

**Comment by dbwiddis** [comment]
OK, the good news... `mvn clean deploy` builds successfully!
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] oshi-parent ........................................ SUCCESS [01:12 min]
[INFO] oshi-core .......................................... SUCCESS [ 45.235 s]
[INFO] oshi-json .......................................... SUCCESS [ 39.692 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:38 min
[INFO] Finished at: 2016-06-22T23:50:09-07:00
[INFO] Final Memory: 87M/1195M
[INFO] ------------------------------------------------------------------------
```
The bad news... still some work to be done!   Here's a list of what I need help with.
 * A look at [the snapshot repository](https://oss.sonatype.org/content/repositories/snapshots/com/github/dblock/) shows it's uploading the `oshi-parent` metadata in addition to the modules.  
    * Is this normal/desired?  
    * Am I going to spam the `com.github.dblock` groupid and should I move things down into an `oshi` directory?  
    * Is this only for the snapshot repository or is the same issue (parent being uploaded) going to occur for releases to the central repository? 
 * I had to make duplicate copies of the `src/assembly` directory contents as well as `LICENSE_HEADER` and `LICENSE_EPL` files.  
    * I see how I could edit each subproject's `assembly.xml` to point to a parent `LICENCE_EPL` but I'm wondering what the standard here is, or if I should make some other change in the `pom.xml` to accomplish that.
    * Will the same thing need to be done for the `site` directory?
 * When this is eventually released, will both subproject site pages be published to the same github site?  Is there additional configuration needed (site.xml?) for that to happen?

**Comment by dbwiddis** [comment]
OK, figured out properties and overriding and got licence and assembly working from the master copy.  Still looking for what I need to do for site to do the same thing.

Also realized that I can't @inheritdoc across the modules so re-copying over interfaces... le sigh.

**Comment by dbwiddis** [comment]
I think I've got the site figured out, thanks to good documentation at wagon-git... which answers the two later questions.   So we're down to groupid and module arrangement.   Questions (And @dblock can chime in here too since it's his github groupid at stake).

1. Do we really need an `oshi-parent`?   Can we not just make `oshi-core` the parent?  That would avoid producing a whole artifact for a parent with no code in it, just metadata.
2. Should we leave  `oshi-core` and `oshi-json` at the root level of the `com.github.dblock` groupID or put them together in an `oshi` subdirectory?  (I was more worried about this with the extra `oshi-parent` as well but if we make `oshi-core` the parent then this may not be as big a deal.)

**Comment by hazendaz** [comment]
Parent no matter the name is how maven works.  If someone grabs core it will pull the metadata pom.  On the site front if using maven site it will bring it all together.  However we could just as well not use that and just use one of the many built in site themes right on github.  The maven generated site for the most part is more developer driven than a casual user may care to see.  Just some ideas anyways...

Get Outlook for Android<https://aka.ms/ghei36>



On Thu, Jun 23, 2016 at 5:08 PM -0400, "dbwiddis" <notifications@github.com<mailto:notifications@github.com>> wrote:


I think I've got the site figured out, thanks to good documentation at wagon-git... which answers the two later questions. So we're down to groupid and module arrangement. Questions (And @dblock<https://github.com/dblock> can chime in here too since it's his github groupid at stake).

  1.  Do we really need an oshi-parent? Can we not just make oshi-core the parent? That would avoid producing a whole artifact for a parent with no code in it, just metadata.
  2.  Should we leave oshi-core and oshi-json at the root level of the com.github.dblock groupID or put them together in an oshi subdirectory? (I was more worried about this with the extra oshi-parent as well but if we make oshi-core the parent then this may not be as big a deal.)

-

**Comment by dbwiddis** [comment]
I just wasted an hour or so trying to make oshi-core the parent.  Unfortunately the requirement to use `<packaging>pom</packaging>` prevents it from working seamlessly and I got tired chasing down all the things that didn't work, so I'm back with an "empty" `oshi-parent`.

I believe it's _probably_ good to go at this point... other than the question of whether to change groupid to a lower level directory (like waffle).

And I'd appreciate a second set of eyes on everything.

**Comment by dbwiddis** [comment]
Seems to be frequently failing its first travis build because the inter-module dependencies aren't being updated in the snapshot repository soon enough.   I've had to manually do `mvn clean deploy` with each update, and then Travis seems to magically work.   I'm sure that means something is broken somewhere but don't know what...  any way to tell the oshi-json project that depends on oshi-core to "use the one you are building here"?

**Comment by hazendaz** [comment]
I'm not seeing why that would occur.  You can try a couple of things like deleting your local .m2/repository location then trying a build to see what occurs.  Travis CI does that.

In the lower poms, remove the relativePath tag.  It's empty which could be causing your issue.  Maven will automatically go to the parent so that is not needed in that case.  If you have that there it needs to say ../pom.xml which is what maven does by default.

I would also avoid using any version of slf4j as done now but instead put a real version there.  Otherwise even your own builds are not necessarly repeatable as things can change and misteriously fail later.  Personally while maven had good intentions with that ability, I find it to be bad from a CI standpoint as it leads to inconsistent builds.

You will want to add the SCM information to each pom file.  Maven inherits that info and if you don't do so, you will end up with issues when you try to release next.  Since this is a single repo, they need to all be the same.  What maven otherwise does is append the model which won't release properly.

**Comment by hazendaz** [comment]
btw - I'm going to be out of action for the next week but promise at some point I will get in here and run some tests and verifications to make sure everything is just right.  You did great work though and I'm sure your maven ninja skills are much better now :)

**Comment by dbwiddis** [comment]
> In the lower poms, remove the relativePath tag. It's empty which could be causing your issue. 

Actually up until my last commit, it was specified as `../pom.xml` which is the default.  That caused my windows VM version of Eclipse to choke so I removed it.   Besides, that's only pointing to the location of the parent, rather than the other module which is the main problem.

> If you have that there it needs to say ../pom.xml which is what maven does by default.

Without it, [I got this error, and fixed it as suggested](http://stackoverflow.com/questions/6003831/parent-relativepath-points-at-my-com-mycompanymyproject-instead-of-org-apache).

> I would also avoid using any version of slf4j as done now but instead put a real version there. 

OK, I'll put it back. I was trying to be flexible.  I'm only using the most basic features of slf4j so I figured I didn't necessarily need to specify the latest patch.

> You will want to add the SCM information to each pom file. Maven inherits that info and if you don't do so, you will end up with issues when you try to release next. Since this is a single repo, they need to all be the same. What maven otherwise does is append the model which won't release properly.

Not sure yet what that means, but I'll research it ... so I need to override the default inherited "same information" and replace it with a specified "same information" so that it doesn't automatically append things?

**Comment by hazendaz** [comment]
If you use eclipse, open the pom and look at the effective pom tab.  It will show you what is occurring.  Also, taking relative path away means that it causes maven to go look into the repository locations as the parent won't be present locally as that is what it's telling it.  The entire tag needs removed.  Most likely that is why you run into build issues on travis.  By default travis doesn't have .m2/repository so everything is 100% new.  The error you get there is telling you exactly that, it could not resolve locally one level higher so it went to the repos.  That in turn means it must have built first.  That further tells me that if its failing there is something wrong but a quick look I don't see any issue.  I'm going to pull this now and give a quick check on it.

**Comment by dbwiddis** [comment]
OK, I think I see.  When I look at the effective pom for `oshi-json` the scm section reads:
```
  <scm>
    <connection>scm:git:git@github.com:dblock/oshi.git/oshi-json</connection>
    <developerConnection>scm:git:git@github.com:dblock/oshi.git/oshi-json</developerConnection>
    <url>https://github.com/dblock/oshi.git/oshi-json</url>
  </scm>
```
But it should just end with `oshi.git`, right?

**Comment by hazendaz** [comment]
It will be identical to the parent with zero changes. See waffe.

BTW this proves the relative path is invalid.  This is me doing this for first time.  I'll submit a PR in a few.

[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for com.github.dblock:oshi-core:[unknown-version]: Could not find artifact com.github.dblock:oshi-parent:pom:3.0-SNAPSHOT and 'parent.relativePath' points at no local POM @ line 4, column 13
[FATAL] Non-resolvable parent POM for com.github.dblock:oshi-json:[unknown-version]: Could not find artifact com.github.dblock:oshi-parent:pom:3.0-SNAPSHOT and 'parent.relativePath' points at no local POM @ line 4, column 13
 @
[ERROR] The build could not read 2 projects -> [Help 1]
[ERROR]
[ERROR]   The project com.github.dblock:oshi-core:[unknown-version] (C:\Users\Jeremy\Documents\GitHub\oshi\oshi-core\pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for com.github.dblock:oshi-core:[unknown-version]: Could not find artifact com.github.dblock:oshi-parent:pom:3.0-SNAPSHOT and 'parent.relativePath' points at no local POM @ line 4, column 13 -> [Help 2]
[ERROR]
[ERROR]   The project com.github.dblock:oshi-json:[unknown-version] (C:\Users\Jeremy\Documents\GitHub\oshi\oshi-json\pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for com.github.dblock:oshi-json:[unknown-version]: Could not find artifact com.github.dblock:oshi-parent:pom:3.0-SNAPSHOT and 'parent.relativePath' points at no local POM @ line 4, column 13 -> [Help 2]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]

**Comment by hazendaz** [comment]
Assembly is in wrong place too.  Create another subproject 'oshi-assembly' to build the assembly.  That is likely also related to issues.  A parent should never build an assembly.  When doing maven each piece should do one thing and one thing well.

**Comment by dbwiddis** [comment]
I honestly don't even know what assembly does!

The error you got above may have been with the empty relativePath which, again, was _not_ present when I had the earlier issues that I reported.

The scm changes are apparently only necessary to build individual projects; if you always build from the parent it apparently doesn't matter.

Regarding your build PR, it won't trigger the issues that I had because the latest OSS snapshot artifact is compatible.  The problem occurs if I make an incompatible change, e.g., refactor a method name (which of course makes it less of a problem once 3.0 is released and the api is consistent).

I'll keep digging around to see what I can figure out.

**Comment by hazendaz** [comment]
My build PR is nothing more than cleanup at this point as it didn't comply properly with maven.  The tidy plugin does that.  You already did the rest.

The error that I showed here was due to not being configured properly to start with and occured on the first build when the parent didn't pre-exist.  The settings you had told it not to look relatively but rather go to the .m2 repo so it was obviously not going to exist unless I had built the parent without any module info in it.  You now are configured properly. 

If you are not using an assembly for anything, then the physical file and execution steps to run it should be removed.  I assumed you were building a zip but after looking at it, it's bulding a bundle with slf4j in it so it's garbage.  You should remove it but leave assembly plugin and version.  That ensures it overrides the hidden maven parent that all project inherit.

SCM is required to perform the release.  It will not work properly if that doesn't adhere.  It does now so you are ok.  When you release, you are going to be releasing the entire set not portions of it.  So it matters.

Between here and at work, I'm the maven master with well over 100 separate projects that I maintain.  Between the changes I suggested and made at the same time (thus I needed to rebase), cleanup of the gitignores, and minor update to site page for the newer site plugin, you are pretty much golden at this point.

I would however merge my PR as it is the run of the tidy command.  Ignore the fact of what the PR looked like at first, I rebased the thing and force pushed it so rather than opening and closing multiple PR, it's just the one and the final contents are the only thing that changed.  And that was basically incorrect formatting and spacing issues between tags.

**Comment by hazendaz** [comment]
Forgot to mention, I'm shutting down now and will be out of state for the next week so you won't hear back from me except maybe via my phone.  I won't be back on github at all.  Take care!

Jeremy

**Comment by dbwiddis** [comment]
The assembly does build a zip (and a tar.bz2 and a tar.gz) containing the jar, api, and all dependency jars (in the target folder of each project), so it does look like it's doing its job as is.  Not sure what we're doing with those... I suppose after a release I could upload the zip files for non-maven download.  

Seems harmless to leave as is for now unless/until we decide what else to do.  Thanks for all the help!

**Comment by dbwiddis** [comment]
Got some help fixing the pom.  The bug I saw was probably associated with restarting the build from scratch because of the way the inheritance worked.  Assembly is now moved into a separate oshi-dist module.  Build is 3x faster and Travis is happy.

**Comment by dbwiddis** [comment]
Annnnnd... almost got it released.  Artifacts are staged just fine but the site plugin isn't happy:

>     [ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.5.1:site (default-site) on project oshi-parent: failed to get report for org.apache.maven.plugins:maven-javadoc-plugin: Failed to execute goal on project oshi-json: Could not resolve dependencies for project com.github.dblock:oshi-json:jar:3.0-rc1: Failure to find com.github.dblock:oshi-core:jar:3.0-rc1 in https://repo.maven.apache.org/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced -> [Help 1]

Possibly relevant discussion [here](https://mail-archives.apache.org/mod_mbox/maven-users/201304.mbox/%3CCAB56zCUv2pm8YjhUxObTDUo5tR4Hj7UCshCP_RxXarHqrZ4itQ@mail.gmail.com%3E).

I will also look into [Maven SCM Publish Plugin](https://maven.apache.org/plugins/maven-scm-publish-plugin/) to replace site plugin.

**Comment by dbwiddis** [comment]
I've spent the better part of the last day, and a few pints of beer, trying to get to the bottom of this build problem.  

The problem occurs in `mvn release:perform` when the goals `deploy site-deploy` are executed together.   Limiting the goal to only 'deploy' works (except that the OSS repository can't be closed to lack of docs) and the site-deploy goal also works by itself for the SNAPSHOT releases (since they are in a repository) but not for any other release.   (Circular dependency: can't do docs without a release and can't release without docs!)

I've finally got past the javadoc issue, I think.  Some combination of the following worked (might try them one at a time to see which):

 - Changed the default goals for `maven-release-plugin` to `install deploy site-deploy`.  (The `install` before the other goals compiles the version being released to the local repository so it's available for the dependency.)
 - Changed the goal for `maven-javadoc-plugin` to `javadoc-no-fork` (apparently the `aggregate` goal doesn't work).

It's now failing on the `oshi-dist` module, and the artifacts uploaded to the central staging repository are missing the sources and javadocs, so they're not releasing there either.  This may be related to me changing the `id` for the javadoc plugin from `attach-javadocs` to `default`... so troubleshooting continues.  But at least progress is being made!

**Comment by dbwiddis** [comment]
I have slain the beast!

3.0-rc1 has been released, but not exactly easily.

 - I had to roll back the attach-sources and attach-javadoc goals to `jar` rather than the `jar-no-fork` options, or they weren't being attached to the OSS release.  This may result in a lot of duplication, and may break the Travis build that I fixed earlier - not sure.  I figured better too many runs than too few...
 - I had to manually install the release to my local repository:
    - After `mvn release:prepare` I rolled back my local repository to the previous commit (the 3.0-rc1 tag) and ran `mvn install` before pulling upstream (to catch back up) to run `mvn release:perform`.

Still somewhat broken:

 - The zip files in the oshi-dist contain only the src/assembly.xml file.   I'm sure that's not what we want... I was looking for the nice zip packages of each module's jar with all its dependencies that we had when it was in the parent.

**Comment by dbwiddis** [comment]
Funny what trial-by-fire does for forcing me to learn more about maven. 😄   (Enjoying your travels, @hazendaz ?)

I think I've got the assembly stuff figured out, will test on next release.   And I just managed to get us on [Coveralls](https://coveralls.io/github/dblock/oshi) which will help me focus my test building for the final 3.0 release. 😈

**Comment by hazendaz** [comment]
Enjoying my travels...glad you are picking up all things maven.  Definately a good toolset to know.  Glad to see coveralls is added.  Keep up the good work!

Get Outlook for Android<https://aka.ms/ghei36>



On Wed, Jun 29, 2016 at 8:25 PM -0400, "dbwiddis" <notifications@github.com<mailto:notifications@github.com>> wrote:


Funny what trial-by-fire does for forcing me to learn more about maven. ? (Enjoying your travels, @hazendaz<https://github.com/hazendaz> ?)

I think I've got the assembly stuff figured out, will test on next release. And I just managed to get us on Coveralls<https://coveralls.io/github/dblock/oshi> which will help me focus my test building for the final 3.0 release. ?

-
